File: dnq-transient-store/src/main/java/com/jetbrains/teamsys/dnq/database/BasePersistentClassImpl.java
/**
 * Copyright 2006 - 2018 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.jetbrains.teamsys.dnq.database;

import jetbrains.exodus.database.TransientEntity;
import jetbrains.exodus.database.TransientEntityStore;
import jetbrains.exodus.database.exceptions.CantRemoveEntityException;
import jetbrains.exodus.database.exceptions.DataIntegrityViolationException;
import jetbrains.exodus.entitystore.Entity;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.*;
import java.util.concurrent.Callable;

public abstract class BasePersistentClassImpl implements Runnable {
    @Nullable
    protected Map<String, Iterable<PropertyConstraint>> propertyConstraints;
    protected TransientEntityStore entityStore;

    protected BasePersistentClassImpl() {
    }

    protected Entity _constructor(@NotNull final String _entityType_) {
        return getEntityStore().getThreadSession().newEntity(_entityType_);
    }

    public boolean isPropertyRequired(String name, Entity entity) {
        return false;
    }

    public TransientEntityStore getEntityStore() {
        return entityStore;
    }

    public void setEntityStore(TransientEntityStore entityStore) {
        this.entityStore = entityStore;
    }

    @NotNull
    public Map<String, Iterable<PropertyConstraint>> getPropertyConstraints() {
        return propertyConstraints != null ? propertyConstraints : Collections.<String, Iterable<PropertyConstraint>>emptyMap();
    }

    protected Map<String, Callable<String>> getPropertyDisplayNames() {
        return null;
    }

    @NotNull
    public String getPropertyDisplayName(String name) {
        final Map<String, Callable<String>> propertyDisplayNames = getPropertyDisplayNames();
        final Callable<String> displayName = propertyDisplayNames == null ? null : propertyDisplayNames.get(name);
        try {
            return displayName == null ? name : displayName.call();
        } catch (Exception ex) {
            throw new RuntimeException(ex);
        }
    }

    public void destructor(Entity entity) {
    }

    public void executeBeforeFlushTrigger(Entity entity) {
    }

    public boolean evaluateSaveHistoryCondition(Entity entity) {
        return false;
    }

    public void saveHistoryCallback(Entity entity) {
    }

    public DataIntegrityViolationException createIncomingLinksException(List<IncomingLinkViolation> linkViolations, final Entity entity) {
        List<Collection<String>> linkDescriptions = new ArrayList<Collection<String>>();
        for (IncomingLinkViolation violation : linkViolations) {
            linkDescriptions.add(violation.getDescription());
        }
        final String displayName = getDisplayName(entity);
        final String displayMessage = "Could not delete " + displayName + ", because it is referenced";
        return new CantRemoveEntityException(entity, displayMessage, displayName, linkDescriptions);
    }

    public IncomingLinkViolation createIncomingLinkViolation(String linkName) {
        return new IncomingLinkViolation(linkName);
    }

    public String getDisplayName(final Entity entity) {
        return toString(entity);
    }

    public String toString(Entity entity) {
        if (entity instanceof TransientEntity) return ((TransientEntity) entity).getDebugPresentation();
        return entity.toString();
    }

    public static <T> Set<T> buildSet(final T[] data) {
        final Set<T> result = new HashSet<T>(data.length);
        for (final T t : data) {
            result.add(t);
        }
        return result;
    }

}

File: dnq-transient-store/src/main/kotlin/com/jetbrains/teamsys/dnq/database/BasePersistentClassImpl.kt
/**
 * Copyright 2006 - 2018 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.jetbrains.teamsys.dnq.database

import jetbrains.exodus.database.TransientEntity
import jetbrains.exodus.database.TransientEntityStore
import jetbrains.exodus.database.exceptions.CantRemoveEntityException
import jetbrains.exodus.database.exceptions.DataIntegrityViolationException
import jetbrains.exodus.entitystore.Entity
import java.util.concurrent.Callable

abstract class BasePersistentClassImpl : Runnable {
    private var _propertyConstraints: MutableMap<String, MutableList<PropertyConstraint<Any?>>>? = null

    val propertyConstraints: Map<String, Iterable<PropertyConstraint<Any?>>>
        get() = _propertyConstraints.orEmpty()

    lateinit var entityStore: TransientEntityStore

    open val propertyDisplayNames: Map<String, Callable<String>>
        get() = emptyMap()

    open protected fun _constructor(_entityType_: String): Entity {
        return entityStore.threadSessionOrThrow.newEntity(_entityType_)
    }

    open fun isPropertyRequired(name: String, entity: Entity): Boolean {
        return false
    }

    open fun getPropertyDisplayName(name: String): String {
        val displayName = this.propertyDisplayNames[name]
        return if (displayName != null) displayName.call() else name
    }

    open fun destructor(entity: Entity) {}

    open fun executeBeforeFlushTrigger(entity: Entity) {}

    @Deprecated("")
    open fun evaluateSaveHistoryCondition(entity: Entity): Boolean {
        return false
    }

    @Deprecated("")
    open fun saveHistoryCallback(entity: Entity) {
    }

    fun <T> addPropertyConstraint(propertyName: String, constraint: PropertyConstraint<T>) {
        val allConstraintsForType = _propertyConstraints ?: LinkedHashMap<String, MutableList<PropertyConstraint<Any?>>>().also { _propertyConstraints = it }
        val constraintsForProperty = allConstraintsForType.getOrPut(propertyName) { ArrayList() }
        @Suppress("UNCHECKED_CAST")
        constraintsForProperty.add(constraint as PropertyConstraint<Any?>)
    }

    fun createIncomingLinksException(linkViolations: List<IncomingLinkViolation>, entity: Entity): DataIntegrityViolationException {
        val linkDescriptions = linkViolations.map { it.description }
        val displayName = getDisplayName(entity)
        val displayMessage = "Could not delete $displayName, because it is referenced"
        return CantRemoveEntityException(entity, displayMessage, displayName, linkDescriptions)
    }

    open fun createIncomingLinkViolation(linkName: String): IncomingLinkViolation {
        return IncomingLinkViolation(linkName)
    }

    open fun getDisplayName(entity: Entity): String {
        return toString(entity)
    }

    fun toString(entity: Entity): String {
        return (entity as? TransientEntity)?.debugPresentation ?: entity.toString()
    }

    companion object {
        @JvmStatic
        fun <T> buildSet(data: Array<T>) = data.toSet()
    }

}
